{
    "title": "How to call idc function with idapython",
    "link": "https://reverseengineering.stackexchange.com/questions/8806/how-to-call-idc-function-with-idapython",
    "content": "I want to script a x86Emu plugin (an IDA emulator) with IdaPython. It's registered idc functions. Below is function names from the plugin source code.\n<pre><code>/* add IDC functions for interacting with the emulator\nEmuRun();\nEmuTrace();\nEmuStepOne();\nEmuTraceOne();\nEmuSync();\nEmuGetReg(regno);\nEmuSetReg(regno, value);\nEmuAddBpt(addr);\n*/\n</code></pre>\nThis functions works well in idc. I found a class calling idc function in <pre><code>idaapi.py</code></pre>.\n<pre><code>class _IdcFunction(object):\n\"\"\"\nInternal class that calls pyw_call_idc_func() with a context\n\"\"\"\ndef \\_\\_init\\_\\_(self, ctxptr):\n    self.ctxptr = ctxptr\n    # Take a reference to the ctypes callback\n    # (note: this will create a circular reference)\n    self.cb   = _IDCFUNC_CB_T(self)\nfp_ptr = property(lambda self: ctypes.cast(self.cb, ctypes.c_void_p).value)\ndef \\_\\_call\\_\\_(self, args, res):\n    return call_idc_func__(self.ctxptr, args, res)\n</code></pre>\nWhen I type this:\n<pre><code>import ctypes\nidaapi._IdcFuntion( ctypes ).\\_\\_call\\_\\_( 'EmuStepOne', ctypes.c_voidp ) \n</code></pre>\nI get the following error:\n<pre><code>ctypes.ArgumentError: argument 1: <type 'exceptions.TypeError'>: Don't know how to convert parameter 1\n</code></pre>\nI don't know why I get this error, has anyone an idea ?\n",
    "votes": "3",
    "answers": 1,
    "views": "2k",
    "tags": [
        "ida",
        "idapython"
    ],
    "user": "names",
    "time": "May 4, 2015 at 15:31",
    "comments": [
        {
            "user": "names",
            "text": "If not need, call the idaapi.run_statements( IDC_SCRIPT_HERE ) it's returned bool\n",
            "time": null
        },
        {
            "user": "tmr232",
            "text": "<pre><code>idc.Eval(\"Some IDC Expression\")</code></pre> is another option.\n",
            "time": null
        }
    ],
    "answers_data": [
        {
            "content": "You get these errors because you're trying to use <pre><code>ctypes</code></pre>, which calls native functions from python using conversions from python objects to low-level C-like types.\nObviously, a python exception is not something ctypes can convert to native representation, so it fails.\nThere's a really simple way to call IDC code from within IDAPython, using <pre><code>idc.Eval</code></pre>, like the following:\n<pre><code>idc.Eval(\"EmuStepOne();\")\n</code></pre>\n",
            "votes": "2",
            "user": "NirIzr",
            "time": "Sep 11, 2016 at 23:57",
            "is_accepted": false,
            "comments": []
        }
    ]
}